File: HwaSrsApi/Models/Card.cs
using System.Collections.Generic;

namespace HwaSrsApi.Models
{
    public class Card
    {
        public int Id { get; set; }

        // TODO remove optional reference (cascade delete hack)
        // write a custom trigger "instead of delete" to handle multipath delete
        // or .OnDelete(DeleteBehavior.ClientCascade)
        public int? CardTypeId { get; set; }
        public virtual CardType CardType { get; set; }

        public int CourseId { get; set; }
        public virtual Course Course { get; set; }

        public CardProgress Progress { get; set; }
        public string ActivationCondition { get; set; }
        public string Tags { get; set; }

        public virtual List<FieldData> Fields { get; set; } = new List<FieldData>();
    }
}

File: src/app/types/card.ts
import { CardProgress } from './card-progress';
import { CardType, Field } from './card-type';
import { Course } from './course';

interface Card {
    id: number;
    cardType: CardType;
    courseId: Course['id'];
    progress?: CardProgress;
    activationCondition?: string;
    tags?: string;
    fields: FieldData[];
}

interface FieldData {
    cardId: Card['id'];
    fieldId: Field['id'];
    value: string;
}

interface NewCard extends Omit<Card, 'id' | 'cardType' | 'fields' | 'progress'> {
    cardTypeId: CardType['id'];
    fields: NewFieldData[];
}

type NewFieldData = Omit<FieldData, 'cardId'>;

export {
    Card,
    FieldData,
    NewCard,
    NewFieldData,
};
